File: framework/python/betterportal/rendering.py
"""Typed HTML callbacks and exact method-local renderer selection."""
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass
import inspect
import re
from typing import Any, Awaitable, Callable, Generic, Iterable, Mapping, TypeVar, cast
from urllib.parse import quote

from .contracts import parse
from .generated_types import RendererDeclaration, RendererDeclarationInput, ViewRenderData, ViewTenantContext, ViewAppContext, ViewRenderDataRequest, ViewRenderDataRoute
from .handler import RequestContext
from .media import NotAcceptable
from .urls import Urls

Result = TypeVar("Result")


@dataclass(frozen=True)
class RenderContext:
    data: ViewRenderData
    urls: Urls

    @property
    def tenant(self) -> ViewTenantContext: return self.data["tenant"]
    @property
    def app(self) -> ViewAppContext: return self.data["app"]
    @property
    def request(self) -> ViewRenderDataRequest: return self.data["request"]
    @property
    def route(self) -> ViewRenderDataRoute: return self.data["route"]

    @classmethod
    def create(cls, request: RequestContext, view_id: str, matched_path: str, renderer: str, mode: str, kind: str,
               key: str | None, status: int, params: Any, query: Any) -> RenderContext:
        route = {"viewId": view_id, "path": matched_path, "renderer": renderer, "mode": mode, "kind": kind, "status": status}
        if key is not None: route["key"] = key
        return cls(parse("ViewRenderDataSchema", {"tenant": request.scope.tenant, "app": request.scope.app,
            "request": {"method": request.method, "path": request.path, "params": params, "query": query}, "route": route}), request.urls)


class Renderer(Generic[Result]):
    def __init__(self, declaration: RendererDeclarationInput, render: Callable[[Result, RenderContext], str | Awaitable[str]]):
        self._declaration: RendererDeclaration = parse("RendererDeclarationSchema", declaration)
        if not callable(render): raise TypeError("A renderer function is required")
        kind, key = self._declaration["kind"], self._declaration.get("key")
        if kind == "page" and key is not None or kind != "page" and key is None:
            raise ValueError("Only fragment/component renderers require a key")
        if key is not None and not re.fullmatch(r"[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+" if kind == "fragment" else r"[A-Za-z0-9_-]+", key):
            raise ValueError("Invalid renderer key")
        self._render = render

    @property
    def declaration(self) -> RendererDeclaration: return deepcopy(self._declaration)

    @property
    def identity(self) -> tuple[str, str, str | None, int]:
        return self._declaration["renderer"], self._declaration["kind"], self._declaration.get("key"), self._declaration["status"]

    async def render(self, data: Result, context: RenderContext) -> str:
        value = self._render(data, context)
        if inspect.isawaitable(value): value = await value
        if not isinstance(value, str): raise TypeError("Renderers must return HTML strings")
        return context.urls.rewrite(value)


def renderers(values: Iterable[Renderer[Result]]) -> tuple[Renderer[Result], ...]:
    result = tuple(values)
    if len({item.identity for item in result}) != len(result): raise ValueError("Duplicate renderer")
    return result


def select(values: Iterable[Renderer[Result]], renderer: str | None, kind: str, key: str | None, status: int = 200) -> Renderer[Result]:
    match = next((item for item in values if item.identity == (renderer, kind, key, status)), None)
    if match is None: raise NotAcceptable("Requested renderer is not available")
    return match


def html_metadata(values: Iterable[Renderer[Any]]) -> dict[str, Any]:
    groups: dict[str, list[Renderer[Any]]] = {}
    for item in values:
        if item.identity[3] == 200: groups.setdefault(item.identity[0], []).append(item)
    result = {}
    for renderer, entries in groups.items():
        variants: list[dict[str, Any]] = []
        modes = []
        for entry in entries:
            _, kind, key, _ = entry.identity
            if kind == "page":
                if "page" not in modes: modes.append("page")
                variants.append({"id": "default", "title": "Default Content", "slotId": "main", "renderModes": ["page", "fragment"]})
            elif kind == "fragment":
                if "fragment" not in modes: modes.append("fragment")
                variants.append({"id": key, "title": key, "slotId": key, "renderModes": ["fragment"]})
        result[renderer] = {"defaultRenderer": "default", "renderModes": modes, "slots": list(dict.fromkeys(item["slotId"] for item in variants)), "renderers": variants}
    return parse("HtmlRepresentationSupportSchema", {"renderers": result})


def content_type(mode: str, chrome: Mapping[str, Any] | None = None) -> str:
    result = "text/html; mode=" + parse("RenderModeSchema", mode)
    for name, value in parse("BetterPortalRouteChromeSchema", chrome or {}).items():
        key = re.sub(r"[_\s]+", "-", re.sub(r"([a-z0-9])([A-Z])", r"\1-\2", name)).lower()
        if not re.fullmatch(r"[a-z][a-z0-9-]*", key): continue
        encoded = quote(value, safe="~()*!.'-") if isinstance(value, str) else str(value).lower()
        result += f"; bp-chrome-{key}={encoded}"
    return result + "; charset=utf-8"

File: framework/dotnet/BetterPortal/Rendering.cs
using System.Text.RegularExpressions;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal;

public sealed record RenderContext(Generated.ViewRenderData Data, CancellationToken Cancellation = default)
{
    public Generated.ViewTenantContext Tenant => Data.Tenant;
    public Generated.ViewAppContext App => Data.App;
    public Generated.ViewRenderDataRequest Request => Data.Request;
    public Generated.ViewRenderDataRoute Route => Data.Route;
    public static RenderContext Create(RequestContext request, string viewId, string matchedPath, string renderer, string mode, string kind,
        string? key, int status, object? @params, object? query, CancellationToken cancellation = default)
    {
        var route = new Node { ["viewId"] = viewId, ["path"] = matchedPath, ["renderer"] = renderer, ["mode"] = mode, ["kind"] = kind, ["status"] = status };
        if (key is not null) route["key"] = key;
        return new(Contracts.Parse<Generated.ViewRenderData>("ViewRenderDataSchema", new Node { ["tenant"] = request.Scope.Tenant, ["app"] = request.Scope.App,
            ["request"] = new Node { ["method"] = request.Method, ["path"] = request.Path, ["params"] = @params, ["query"] = query }, ["route"] = route }), cancellation);
    }
}

public abstract class Renderer
{
    public Generated.RendererDeclaration Declaration { get; }
    public (string Renderer, string Kind, string? Key, int Status) Identity { get; }
    protected Renderer(Generated.RendererDeclarationInput declaration)
    {
        Declaration = Contracts.Parse<Generated.RendererDeclaration>("RendererDeclarationSchema", declaration);
        var kind = Declaration.Kind.ToString().ToLowerInvariant();
        var key = Declaration.Key.HasValue ? Declaration.Key.Value : null;
        if (kind == "page" && key is not null || kind != "page" && key is null) throw new ArgumentException("Only fragment/component renderers require a key");
        if (key is not null && !Regex.IsMatch(key, kind == "fragment" ? @"\A[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\z" : @"\A[A-Za-z0-9_-]+\z"))
            throw new ArgumentException("Invalid renderer key");
        Identity = (Declaration.Renderer, kind, key, (int)Declaration.Status);
    }
    internal abstract ValueTask<string> RenderBoxed(object? data, RenderContext context);
    public ValueTask<string> RenderUntyped(object? data, RenderContext context) => RenderBoxed(data, context);
    internal static IReadOnlyList<T> Unique<T>(IEnumerable<T> values) where T : Renderer
    {
        var result = values.ToArray();
        if (result.Select(item => item.Identity).Distinct().Count() != result.Length) throw new ArgumentException("Duplicate renderer");
        return Array.AsReadOnly(result);
    }
    public static T Select<T>(IEnumerable<T> values, string? renderer, string kind, string? key = null, int status = 200) where T : Renderer =>
        values.FirstOrDefault(item => item.Identity == (renderer, kind, key, status)) ?? throw new NotAcceptableException("Requested renderer is not available");
    public static string ContentType(string mode, object? chrome = null)
    {
        var result = "text/html; mode=" + Contracts.Parse("RenderModeSchema", mode);
        foreach (var (name, value) in (Node)Contracts.Parse("BetterPortalRouteChromeSchema", chrome ?? new Node())!)
        {
            var key = Regex.Replace(Regex.Replace(name, "([a-z0-9])([A-Z])", "$1-$2"), @"[_\s]+", "-").ToLowerInvariant();
            if (!Regex.IsMatch(key, @"\A[a-z][a-z0-9-]*\z")) continue;
            var encoded = value is string text ? Uri.EscapeDataString(text).Replace("%21", "!").Replace("%27", "'").Replace("%28", "(").Replace("%29", ")").Replace("%2A", "*")
                : Convert.ToString(value, System.Globalization.CultureInfo.InvariantCulture)!.ToLowerInvariant();
            result += "; bp-chrome-" + key + "=" + encoded;
        }
        return result + "; charset=utf-8";
    }
    internal static Node HtmlMetadata(IEnumerable<Renderer> values)
    {
        var result = new Node();
        foreach (var group in values.Where(item => item.Identity.Status == 200).GroupBy(item => item.Identity.Renderer))
        {
            var modes = new List<string>(); var variants = new List<Node>();
            foreach (var entry in group)
            {
                if (entry.Identity.Kind == "page")
                {
                    modes.Add("page");
                    variants.Add(new() { ["id"] = "default", ["title"] = "Default Content", ["slotId"] = "main", ["renderModes"] = new[] { "page", "fragment" } });
                }
                else if (entry.Identity.Kind == "fragment")
                {
                    modes.Add("fragment");
                    variants.Add(new() { ["id"] = entry.Identity.Key, ["title"] = entry.Identity.Key, ["slotId"] = entry.Identity.Key, ["renderModes"] = new[] { "fragment" } });
                }
            }
            result[group.Key] = new Node { ["defaultRenderer"] = "default", ["renderModes"] = modes.Distinct().ToArray(),
                ["slots"] = variants.Select(item => item["slotId"]).Distinct().ToArray(), ["renderers"] = variants };
        }
        return (Node)Contracts.Parse("HtmlRepresentationSupportSchema", new Node { ["renderers"] = result })!;
    }
}

public sealed class Renderer<TResult> : Renderer
{
    private readonly Func<TResult, RenderContext, ValueTask<string>> render;
    public Renderer(Generated.RendererDeclarationInput declaration, Func<TResult, RenderContext, ValueTask<string>> render) : base(declaration)
    { ArgumentNullException.ThrowIfNull(render); this.render = render; }
    public Renderer(Generated.RendererDeclarationInput declaration, Func<TResult, RenderContext, string> render) : this(declaration, (data, context) => ValueTask.FromResult(render(data, context)))
    { ArgumentNullException.ThrowIfNull(render); }
    public async ValueTask<string> Render(TResult data, RenderContext context)
    {
        context.Cancellation.ThrowIfCancellationRequested();
        return await render(data, context).AsTask().WaitAsync(context.Cancellation) ?? throw new InvalidOperationException("Renderers must return HTML strings");
    }
    internal override ValueTask<string> RenderBoxed(object? data, RenderContext context) => Render((TResult)data!, context);
}
